### ISSUE TITLE
Add Symbol or less restrictive Path

### ISSUE THREAD

**Issue by gtnicol** [immutables/immutables#1256]
While having Path be typed and bound to the object hierarchy is nice, it also make query rewriting very difficult. I often have the need to convert from one representation, like 'foo==1' into something like 'qux.bar.baz==1' which *can* be handled by different objects it's a nuisance, especially if trying to convert to a reference deep in the object graph.

I also have most of a SQL criteria written... are you interested in a PR?

**Comment by asereda-gs** [comment]
It is true that right now Path depends on reflection (`java.lang.reflect.Member`) probably that can be relaxed.
You can still rewrite AST using Visitors.
Can you please give more details on your usecase ?

**Comment by gtnicol** [comment]
The general use case is to use Query as the canonical model of a query in an application. I do pretty extensive rewriting via Visitors and one part of this is the need to rename things, especially where the backend database structure/names aren't exactly the same structure as the object graph (i.e. where there is no direct mapping from a Member to a backend database structure or when mapping to an element in a collection). In my case I'd like to map from a ReST query in RSQL or something like that into a Query against the models exposed the the ReST endpoint, and then rewrite that query against a backend store (sql, mongo, BigQuery, etc). Having something like a Symbol or a relaxed Path would make it easier (do need to hack around by introducing fake objects). Making the Visitor more easily extensible to Expression subtypes might be another thing.

**Comment by asereda-gs** [comment]
So, if I understand you correctly, you would like the following:

```text
RSQL -> Criteria AST -> execute/obtain Native Query (mongo, SQL, elastic etc.)
```

This can be based on a generic model, not necessarily existing java class (reflection) ?

**Comment by gtnicol** [comment]
That's right. I already have working RSQL -> Criteria AST -> (mongo, SQL, BQ) and have mapping between layers (DTO<->internal<->ReST), so the queries get rewritten between layers to support the final mapping (using visitors) to native query.

In order to create new Path objects during the rewrite I have to introduce 'fake' objects with a field or method with the name I want so I can call Path.ofMember() using the result of Reflections.findMember()... which is a bit of a hack. Path doesn't check the validity of parent/child anyway.

FWIW. RSQL isn't the requirement... I could (and probably will) support something else.

**Comment by asereda-gs** [comment]
The reason Path currently requires `java.lang.reflect.Member` to operate is because backends might need type / annotation metadata when constructing or fetching the queries.  Unfortunately right now we don't have a generic Model API to replace reflection.

Probably with SQL backend we might address it.

**Comment by gtnicol** [comment]
Sure, and that makes sense when there is a direct mapping to the underlying store (where the objects are kind of like straight DTO). I guess this use-case is a little outside of that realm because it's a mapping from an external representation to an internal one (from an external language across an internal model to multiple stores).

That said, it's often useful to have custom mappers to/from the underlying database, in which case you can run into the same kind of problem. In the SQL backend I have, I allow mapping functions to be passed down via config, and at the lower layers, the system used entityType on the Query to find the mapping function to covert from row->object.
